## **Scenario A**

In [1]:
import itertools
import numpy as np

def prob_sum_greater_than(m: int, n: int, x: int):
  """
  This function calculates the probabilty of the sum of 'm' dice with 'n' sides
  each to be greater than the target sum 'x'.

  Arguments:
  m -- number of dice
  n -- number of sides on each die
  x -- target sum

  Returns:
  prob -- probability
  """
  # Validate input parameters
  if not (1 <= m <= 4):
      raise ValueError("Number of dice 'm' must be between 1 and 4.")
  if not (1 <= n <= 20):
      raise ValueError("Number of sides 'n' must be 20 or less.")

  # Generate all possible roll combinations
  rolls = list(itertools.product(range(1, n + 1), repeat=m))

  # Calculate the sum of each roll
  sums = np.array([sum(roll) for roll in rolls])

  # Calculate the probability
  prob = np.mean(sums > x)

  return prob


## **Scenario** **B**

In [2]:
def prob_one_die_greater_than(m: int, n: int, x: int):
  """
  This function calculates the probabilty of rolling at least one die with
  a value greater than or equal to 'x' when rolling 'm' dice with 'n' sides.

  Arguments:
  m -- number of dice
  n -- number of sides on each die
  x -- target number

  Returns:
  prob -- probability
  """
  # Calculate the probability that a single die fails to meet the target
  fail_probability = (x - 1) / n

  # Probability that all dice fail to meet the target
  all_fail_probability = fail_probability ** m

  # Probability that at least one die meets or exceeds the target
  prob = 1 - all_fail_probability

  return prob


# **Situation** **1**

In [4]:
m1 = 1
n1 = 20
x1 = 16  #since the function A calculates greater than not greater than or equal to.

prob_challenge1 = prob_sum_greater_than(m1, n1, x1)
print(f"The probability of rolling a sum of {x1 + 1} or more when rolling {m1} dice with {n1} sides is: {prob_challenge1}")

The probability of rolling a sum of 17 or more when rolling 1 dice with 20 sides is: 0.2


# **Situation** **2**

In [5]:
m2 = 3
n2 = 4
x2 = 9  #since the function A calcualtes greater than not greater than or equal to.

prob_challenge2 = prob_sum_greater_than(m2, n2, x2)
print(f"The probability of rolling a sum of {x2 + 1} or more when rolling {m2} dice with {n2} sides is: {prob_challenge2}")

The probability of rolling a sum of 10 or more when rolling 3 dice with 4 sides is: 0.15625


# **Situation** **3**

In [6]:
m3 = 4
n3 = 6
x3 = 6

prob_challenge3 = prob_one_die_greater_than(m3, n3, x3)
print(f"The probability of rolling at least one die with a value greater than or equal to {x3} when rolling {m3} dice with {n3} sides is: {prob_challenge3}")

The probability of rolling at least one die with a value greater than or equal to 6 when rolling 4 dice with 6 sides is: 0.5177469135802468


# **Conclusion**
From all these probabilities found, it is clear that situation C has the best odds of going in favor of the player with about a 52% chance of the player being successful. The next best option would be Situation A with a 20% chance of success. Finally, the worst option would be situation B where the chance of success is only about 16%.